### PR TITLE
Query: Read Avg/Min/Max as nullable from database to throw exception …

### LINKED ISSUE #18955: Max with 0 result give error Sequence contains no elements when using .Max() or .MaxAsync()
When the result of max is 0 a error message is generated

### Steps to reproduce

``` C#
_ctx.ChapterQuestions.Where(x => x.SurveyChapterId == surveyChapterId).Max(x => x.SequenceNumber+0) // produce error

_ctx.ChapterQuestions.Where(x => x.SurveyChapterId == surveyChapterId).Max(x => x.SequenceNumber+1) // is allowed
```
sql data
```
SELECT id,SequenceNumber FROM [ChapterQuestions] AS [c] WHERE [c].[SurveyChapterId] = 702
---
Id	SequenceNumber	
2293	0	
```
stack trace
```
dbug: Microsoft.EntityFrameworkCore.Query[10107]
      (queryContext) => Single<int>(new QueryingEnumerable<int>(
          (RelationalQueryContext)queryContext,
          SqlServerQuerySqlGeneratorFactory,
          SqlExpressionFactory,
          ParameterNameGeneratorFactory,
          Projection Mapping:
              EmptyProjectionMember -> 0
          SELECT MAX(c.SequenceNumber + 0)
          FROM ChapterQuestions AS c
          WHERE (c.SurveyChapterId == @__surveyChapterId_0) && NotEqual(@__surveyChapterId_0),
          Func<QueryContext, DbDataReader, ResultContext, int[], ResultCoordinator, int>,
          Survey.Entities.Context.SurveyContext,
          DiagnosticsLogger<Query>
      ))
dbug: Microsoft.EntityFrameworkCore.Database.Command[20103]
      Creating DbCommand for 'ExecuteReader'.
dbug: Microsoft.EntityFrameworkCore.Database.Command[20104]
      Created DbCommand for 'ExecuteReader' (0ms).
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20000]
      Opening connection to database 'survey' on server '(localdb)\mssqllocaldb'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20001]
      Opened connection to database 'survey' on server '(localdb)\mssqllocaldb'.
info: Microsoft.EntityFrameworkCore.Database.Command[20100]
      Executing DbCommand [Parameters=[@__surveyChapterId_0='702'], CommandType='Text', CommandTimeout='30']
      SELECT MAX([c].[SequenceNumber] + 0)
      FROM [ChapterQuestions] AS [c]
      WHERE ([c].[SurveyChapterId] = @__surveyChapterId_0) AND @__surveyChapterId_0 IS NOT NULL
dbug: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (3ms) [Parameters=[@__surveyChapterId_0='702'], CommandType='Text', CommandTimeout='30']
      SELECT MAX([c].[SequenceNumber] + 0)
      FROM [ChapterQuestions] AS [c]
      WHERE ([c].[SurveyChapterId] = @__surveyChapterId_0) AND @__surveyChapterId_0 IS NOT NULL
fail: Microsoft.EntityFrameworkCore.Query[10100]
      An exception occurred while iterating over the results of a query for context type 'Survey.Entities.Context.SurveyContext'.
      System.InvalidOperationException: Sequence contains no elements.
         at lambda_method(Closure , QueryContext , DbDataReader , ResultContext , Int32[] , ResultCoordinator )
         at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
System.InvalidOperationException: Sequence contains no elements.
   at lambda_method(Closure , QueryContext , DbDataReader , ResultContext , Int32[] , ResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
dbug: Microsoft.EntityFrameworkCore.Database.Command[20300]
      A data reader was disposed.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20002]
      Closing connection to database 'survey' on server '(localdb)\mssqllocaldb'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20003]
      Closed connection to database 'survey' on server '(localdb)\mssqllocaldb'.
info: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[2]
      Executed action Survey.WebApi.Survey.ManageSurveys.Controllers.SurveyManagementController.InsertChapterQuestion (Survey.WebApi) in 39.4262ms
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'Survey.WebApi.Survey.ManageSurveys.Controllers.SurveyManagementController.InsertChapterQuestion (Survey.WebApi)'
fail: Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware[1]
      An unhandled exception has occurred while executing the request.
System.InvalidOperationException: Sequence contains no elements.
   at lambda_method(Closure , QueryContext , DbDataReader , ResultContext , Int32[] , ResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.Max[TSource,TResult](IQueryable`1 source, Expression`1 selector)
   at Survey.WebApi.Survey.ManageSurveys.DbServices.ManageSurveyQuestionProvider.InsertChapterQuestion(Int32 surveyId, Int32 surveyChapterId, Int32 questionId) in C:\git\mijnzorgpeiler\Survey\Survey.WebApi\Survey\ManageSurveys\DbServices\ManageSurveyQuestionProvider.cs:line 50
   at Survey.WebApi.Survey.ManageSurveys.Services.ManageSurveyService.InsertChapterQuestion(Int32 surveyId, Int32 surveyChapterId, Int32 questionId) in C:\git\mijnzorgpeiler\Survey\Survey.WebApi\Survey\ManageSurveys\Services\ManageSurveyService.cs:line 69
   at Survey.WebApi.Survey.ManageSurveys.Controllers.SurveyManagementController.InsertChapterQuestion(Int32 surveyId, Int32 surveyChapterId, Int32 questionId) in C:\git\mijnzorgpeiler\Survey\Survey.WebApi\Survey\ManageSurveys\Controllers\SurveyManagementController.cs:line 59
   at lambda_method(Closure , Object )
   at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
```

### Further technical details

EF Core version:
Database provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer)
Target framework: (e.g. .NET Core 3.1)
Operating system:
IDE: (e.g. Visual Studio 2019 16.4)

### PR DESCRIPTION
**Pull request by smitpatel**
…only when it is empty

Resolves #18955

### CODE REVIEW

**Comment by smitpatel** [comment]
Description
Since 3.0, we read value from database according to the type. In Avg/Min/Max database returns null if there are no elements. We read this as default value of non-nullable type and throw exception if we hit default. If the actual result of above methods ends up being same as default value, we incorrect identify it is empty sequence.

Customer Impact
Query failure when result of Avg/Min/Max in database is 0.

How found
Customer-reported (#18955)

Test coverage
We did not have tests when Avg/Min/Max returned 0 result.

Regression?
Yes, from 2.2.

Risk
Risk is very small - the fix is reading database value as nullable for those functions and disambiguate null from 0 in result to throw exception only in correct case.

**Comment by ajcvickers** [comment]
@Pilchie FYI, we would like to patch this in 3.1.1.

**Comment by Pilchie** [comment]
What workaround is available to the customer, if any?

**Comment by ajcvickers** [comment]
@Pilchie Having a workaround may not help much in this case. The issue is that if you are calculating the max of a set of values, then you're application may work fine _until the result happens to be zero_ at which point the app will crash. To use a workaround you would have to know that the issue exists, which is very easily missed here, even with reasonable testing.

File: src/EFCore.Relational/Storage/FloatTypeMapping.cs [hunk 1]
```diff
@@ -55,6 +56,10 @@ protected override RelationalTypeMapping Clone(RelationalTypeMappingParameters p
         ///     The generated string.
         /// </returns>
         protected override string GenerateNonNullSqlLiteral(object value)
-            => ((float)value).ToString("R", CultureInfo.InvariantCulture);
+        {
+            var floatValue = Convert.ToSingle(value);
```
Comment (maumar) [inline src/EFCore.Relational/Storage/FloatTypeMapping.cs]: keep as lambda body

**Comment by ajcvickers** [comment]
@Pilchie @wtgodbe @dougbu Can this now be merged for 3.1.1?

**Pull request merged**